### PR TITLE
Feature/cdap 14770 migrate preferences dataset

### PR DESCRIPTION
**Pull request by jayfjin**
Migrate PreferencesDataset to the new SPI. Refactor DefaultConfigStore to use a DefaultConfigTable, since PreferencesDataset does not control the transaction itself.

Build: https://builds.cask.co/browse/CDAP-DUT6848-1
https://builds.cask.co/browse/IT-ITM-108

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesTable.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        default:
+          throw new UnsupportedOperationException(
+            String.format("Preferences cannot be used on this entity type: %s", entityId.getEntityType()));
+      }
+    } catch (ConfigNotFoundException e) {
+      //no-op
+    }
+  }
+
+  private String getProgramName(ProgramId programId) {
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesTable.java:195]: Is this method here because program can same name but in different app or type? If so, it is better to add a comment why program is treated different than app and namespace.

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/DefaultConfigStore.java [hunk 2]
```diff
@@ -44,8 +31,6 @@
  * "
  */
 public class DefaultConfigStore implements ConfigStore {
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/main/java/co/cask/cdap/config/DefaultConfigStore.java:33]: I think we can remove the ConfigStore interface and just have this class named ConfigStore. The interface only have 1 implementation and it adds some complexity in guice injection.

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/DefaultConfigTable.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Table methods for {@link DefaultConfigStore}
+ */
+public class DefaultConfigTable {
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/main/java/co/cask/cdap/config/DefaultConfigTable.java]: If we remove the ConfigStore interface, this class can be renamed to ConfigTable

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataSubscriberService.java [hunk 4]
```diff
@@ -186,7 +187,8 @@ protected boolean shouldRunInSeparateTx(MetadataMessage message) {
 
   @Override
   protected void processMessages(DatasetContext datasetContext, StructuredTableContext structuredTableContext,
-                                 Iterator<ImmutablePair<String, MetadataMessage>> messages) throws IOException {
+                                 Iterator<ImmutablePair<String, MetadataMessage>> messages)
+    throws IOException, ConfigNotFoundException {
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataSubscriberService.java]: Unnecessary change?

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/DefaultConfigTable.java (hunk 3 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/config/DefaultConfigTable.java]: This table is also used by the PreferencesTable, better to add javadoc

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesTable.java (hunk 1 above)
Comment (jayfjin) [reply cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesTable.java:195]: Done

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/DefaultConfigStore.java (hunk 2 above)
Comment (jayfjin) [reply cdap-app-fabric/src/main/java/co/cask/cdap/config/DefaultConfigStore.java:33]: I'd rather keep this as is - this seems like a common pattern throughout the CDAP codebase. If this is something we wish to change we should file a JIRA.

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/DefaultConfigTable.java (hunk 3 above)
Comment (jayfjin) [reply cdap-app-fabric/src/main/java/co/cask/cdap/config/DefaultConfigTable.java]: Done

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataSubscriberService.java (hunk 4 above)
Comment (jayfjin) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataSubscriberService.java]: Thanks, done

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/DefaultConfigStore.java (hunk 2 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/config/DefaultConfigStore.java:33]: yea some of them are useful when we need to have different implementations in some other environment or unit test, I think you can file a JIRA to remove the interface. I think in this pr you can just rename `DefaultConfigTable` to `ConfigTable` since it is a newly added class

Comment (jayfjin) [reply cdap-app-fabric/src/main/java/co/cask/cdap/config/DefaultConfigStore.java:33]: renamed

**Review by yaojiefeng — APPROVED** [review]
lgtm
